### PR TITLE
Share TFBuffer between functions remove Mutex

### PR DESCRIPTION
**Pull request by MaxiMaerz**
Closes: 19

### CODE REVIEW

File: src/app.rs [hunk 1]
```diff
@@ -25,9 +23,15 @@ pub enum AppModes{
...
-            ref_transform: &Arc<RwLock<rosrust_msg::geometry_msgs::Transform>>, axis_length: f64
+            tf_listener: Arc<rustros_tf::TfListener>,
+            static_frame: String,
+            axis_length: f64
         ) -> Vec<Line> {
-        let tf = &ref_transform.as_ref().read().unwrap();
+        let tf = tf_listener.lookup_transform(
+            &static_frame,
+            "base_link",
+            rosrust::Time::new()
```
Comment (carzum) [inline src/app.rs:33]: is this. ros::Time(0) in rosrust?

File: src/app.rs [hunk 2]
```diff
@@ -91,26 +101,45 @@ impl App{
...
+            tf_listener: Arc<rustros_tf::TfListener>,
+        )
         {
             // 0.5 is the height width ratio of terminal chars
             let scale_factor = self.terminal_size.0 as f64 / self.terminal_size.1 as f64 * 0.5;
-            let tf = &ref_transform.as_ref().read().unwrap();
+            let listener = tf_listener.clone();
+            let res = &listener.clone().lookup_transform(
+                &self.static_frame,
+                "base_link",
```
Comment (carzum) [inline src/app.rs:129]: This could also be part of the config, it would allow custom camera frames

File: src/app.rs [hunk 3]
```diff
@@ -91,26 +101,45 @@ impl App{
...
+                rosrust::Time::new()
+            );
+            match res {
+                Ok(res) => res,
+                Err(_e) => return,
```
Comment (carzum) [inline src/app.rs:134]: can we deal with this?

File: src/app.rs [hunk 4]
```diff
@@ -153,7 +183,9 @@ impl App{
                         });
                     }
                     ctx.layer();
-                    for elem in App::calculate_footprint(tf) {
+                    for elem in App::calculate_footprint(
```
Comment (carzum) [inline src/app.rs:186]: We are generating a whole bunch of tf lookups here

File: src/app.rs [hunk 5]
```diff
@@ -173,7 +205,10 @@ impl App{
                             ctx.draw(&line);
                         };
                     }
-                    for line in get_frame_lines(tf, self.axis_length) { ctx.draw(&line); };
+                    for line in get_frame_lines(
```
Comment (carzum) [inline src/app.rs:208]: same here

File: src/main.rs [hunk 6]
```diff
@@ -29,55 +27,43 @@ fn main() -> Result<(), Box<dyn Error>> {
...
-                let mut cb_tf = cb_tf.write().unwrap();
-                *cb_tf = res.as_ref().unwrap().transform.clone();
-                }
-            }
-            thread::sleep(Duration::from_millis(sleep));
-        }
-    });
+    // Initialize listener and wait for it to come up
+    let listener = Arc::new(rustros_tf::TfListener::new());
+    while rosrust::is_ok() {
```
Comment (carzum) [inline src/main.rs:33]: this helps 👍

Comment (carzum) [reply src/main.rs:33]: wait forever is ok?

**Review by carzum — COMMENTED** [review]
I think the amount of tf lookups might have increased with this commit, which would add more 'sleeps' to the code. @MaxiMaerz  did you see any performance differences / can we even measure that?

File: src/main.rs (hunk 6 above)
Comment (MaxiMaerz) [reply src/main.rs:33]: This just waits for the buffer to fill,
I can print something but think its fine.

File: src/app.rs (hunk 4 above)
Comment (Eman7C7) [reply src/app.rs:186]: why? we are iterating on a vector! I doubt calculate_footprint is called at every iteration of the for... Riiiight?

**Review by Eman7C7 — CHANGES_REQUESTED** [review]

File: src/app.rs (hunk 4 above)
Comment (carzum) [reply src/app.rs:186]: you are right, my bad

File: src/app.rs (hunk 1 above)
Comment (MaxiMaerz) [reply src/app.rs:33]: Time { sec: 0, nsec: 0 }
I would say so

File: src/app.rs (hunk 3 above)
Comment (MaxiMaerz) [reply src/app.rs:137]: I think its ok we just try next frame

**Review by Eman7C7 — APPROVED** [review]

**Pull request merged**